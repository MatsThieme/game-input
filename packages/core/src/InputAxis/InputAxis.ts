type NumberTupleMinLength = [number, ...number[]];

export class InputAxis {
    private _values: NumberTupleMinLength;
    private _length?: number;
    private _changed: number;

    public constructor(values?: number | NumberTupleMinLength) {
        if (values === undefined) this._values = [0];
        else if (typeof values === "number") this._values = [values];
        else this._values = values.slice() as NumberTupleMinLength;

        this._changed = 0;
    }

    /**
     * Returns whether values changed in the last frame.
     */
    public get changed(): boolean {
        return this._changed === 1;
    }

    public getValues(): Readonly<NumberTupleMinLength> {
        return this._values;
    }

    /**
     * @internal
     */
    public setValues(values: Readonly<NumberTupleMinLength>): void {
        if (
            this._values.length !== values.length ||
            this._values.some((value, index) => value !== values[index])
        ) {
            this._values = values.slice() as NumberTupleMinLength;

            this._length !== undefined && (this._length = undefined);
            this._changed = 2;
        }
    }

    /**
     * @internal
     */
    public update(): void {
        this._changed > 0 && this._changed--;
    }

    public getLength(): number {
        if (typeof this._length === "number") {
            return this._length;
        }

        if (this._values.length === 1) {
            return (this._length = this._values[0]);
        }

        return (this._length = Math.sqrt(
            this._values.reduce((result, current) => result + current ** 2, 0)
        ));
    }
}
