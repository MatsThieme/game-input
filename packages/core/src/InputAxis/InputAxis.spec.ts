import { InputAxis } from "./InputAxis";
import { describe, it, expect } from "vitest";

describe("InputAxis", () => {
    it.each([
        [new InputAxis()],
        [new InputAxis(0)],
        [new InputAxis([0])],
        [new InputAxis([0, 0])],
        [new InputAxis(1)],
        [new InputAxis([1])],
        [new InputAxis([1, 0])],
        [new InputAxis([0, 1])],
        [new InputAxis([0.7, 0.7])],
        [new InputAxis([0.7, 0.7, 1, 1, 1, 1])],
    ])("should calculate the length of its values", (axis: InputAxis) => {
        snapAxis(axis);
    });

    it("should update correctly", () => {
        const axis = new InputAxis();

        snapAxis(axis);

        axis.setValues([1, 2, 3]);

        snapAxis(axis);

        axis.update();

        snapAxis(axis);

        axis.update();

        snapAxis(axis);
    });
});

function snapAxis(axis: InputAxis): void {
    expect({
        changed: axis.changed,
        length: axis.getLength(),
        values: axis.getValues(),
    }).toMatchSnapshot();
}
