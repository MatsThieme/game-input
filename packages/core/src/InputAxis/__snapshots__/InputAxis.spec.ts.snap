// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputAxis > should calculate the length of its values 1`] = `
{
  "changed": false,
  "length": 0,
  "values": [
    0,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 2`] = `
{
  "changed": false,
  "length": 0,
  "values": [
    0,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 3`] = `
{
  "changed": false,
  "length": 0,
  "values": [
    0,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 4`] = `
{
  "changed": false,
  "length": 0,
  "values": [
    0,
    0,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 5`] = `
{
  "changed": false,
  "length": 1,
  "values": [
    1,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 6`] = `
{
  "changed": false,
  "length": 1,
  "values": [
    1,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 7`] = `
{
  "changed": false,
  "length": 1,
  "values": [
    1,
    0,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 8`] = `
{
  "changed": false,
  "length": 1,
  "values": [
    0,
    1,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 9`] = `
{
  "changed": false,
  "length": 0.9899494936611665,
  "values": [
    0.7,
    0.7,
  ],
}
`;

exports[`InputAxis > should calculate the length of its values 10`] = `
{
  "changed": false,
  "length": 2.23159136044214,
  "values": [
    0.7,
    0.7,
    1,
    1,
    1,
    1,
  ],
}
`;

exports[`InputAxis > should update correctly 1`] = `
{
  "changed": false,
  "length": 0,
  "values": [
    0,
  ],
}
`;

exports[`InputAxis > should update correctly 2`] = `
{
  "changed": false,
  "length": 3.7416573867739413,
  "values": [
    1,
    2,
    3,
  ],
}
`;

exports[`InputAxis > should update correctly 3`] = `
{
  "changed": true,
  "length": 3.7416573867739413,
  "values": [
    1,
    2,
    3,
  ],
}
`;

exports[`InputAxis > should update correctly 4`] = `
{
  "changed": false,
  "length": 3.7416573867739413,
  "values": [
    1,
    2,
    3,
  ],
}
`;
